lib/cadet/accounts/luminus.ex

```diff
@@ -0,0 +1,211 @@
+    - {:error, :bad_request} - invalid token
+    - {:error, :internal_server_error} - the luminus_client_secret is invalid
+
+  ## Parameters
+
+    - code: String, the LumiNUS authorization code
+
+  ## Examples
+
+      iex> Cadet.Accounts.LumiNUS.fetch_luminus_token("C0dE...")
+      {:ok, "12742174091894830298409823098098"}
+  """
+
+  def fetch_luminus_token(code) do
+    queries =
+      %{
+        client_id: @client_id,
+        client_secret: @client_secret,
+        grant_type: "authorization_code",
+        code: code,
+        redirect_uri: @redirect_url
+      }
+      |> URI.encode_query()
+
+    headers = [{"Content-type", "application/x-www-form-urlencoded"}]
+
+    case HTTPoison.post(@api_token_url, queries, headers) do
+      {:ok, %{body: body, status_code: 200}} when body != ~s("") ->
+        {:ok, body |> Jason.decode!() |> Map.get("access_token")}
+
+      {:ok, %{status_code: 400}} ->
+        # LumiNUS responds with 400 if APIKey is invalid
+        {:error, :bad_request}
+    end
+  end
+
+  @doc """
+  Get the NUSNET ID and the name of the user corresponding to this token.
+
+  returns...
+
+    - {:ok, nusnet_id, name} - valid token, nusnet_id are name are strings
+    - {:error, :bad_request} - invalid token
+    - {:error, :internal_server_error} - the luminus_client_secret is invalid
+
+  ## Parameters
+
+    - token: String, the LumiNUS authentication token
+
+  ## Examples
+
+      iex> Cadet.Accounts.LumiNUS.fetch_details("T0K3N...")
+      {:ok, "e012345", "LIOW JIA CHEN"}
+
+  """
+  def fetch_details(token) do
+    case api_call("user/Profile", Token: token) do
+      {:ok, %{"userID" => nusnet_id, "userNameOriginal" => name}} -> {:ok, nusnet_id, name}
+      _ -> {:error, :bad_request}
+    end
+  end
+
+  @doc """
+  Get the role of the user corresponding to this token.
+
+  returns...
+
+    - {:ok, :student} - valid token, access_Full set to true
+    - {:ok, :admin} - valid token, access_Create set to true
+    - {:ok, :staff} - valid token, access_Read set to true
+    - {:error, :bad_request} - invalid token, or not taking the module
+    - {:error, :internal_server_error} - the lumiNUS_client_secret is invalid
+
+  ## Parameters
+
+    - token: String, the LumiNUS authentication token
+
+  This function assumes that inactive modules have an ID of
+  `"00000000-0000-0000-0000-000000000000"`, and that there is only one active
+  module with the course code `"CS1101S"`. (So far, these assumptions have been
+  true).
+
+  (^need to check if this is still a thing in lumiNUS)
+
+  ## Parameters
+
+    - token: String, the LumiNUS authentication token
+
+  ## Examples
+
+      iex> Cadet.Accounts.Luminus.fetch_role("T0K3N...")
+      {:ok, :student}
+  """
+
+  def fetch_role(token) do
+    case api_call("module", Token: token) do
+      {:ok, modules} ->
+        parse_modules(modules)
+
+      {:error, _} ->
+        {:error, :bad_request}
+    end
+  end
+
+  defp parse_cs1101s(cs1101s) do
```

Comment:
```suggestion
  defp parse_cs1101s({:ok, %{"access" => %{"access_Full" => true}}), do: {:ok, :admin}
  defp parse_cs1101s({:ok, %{"access" => %{"access_Create" => true}}), do: {:ok, :staff}
  defp parse_cs1101s({:ok, %{"access" => %{"access_Read" => true}}), do: {:ok, :student}
  defp parse_cs1101s(_), do: {:error, :bad_request}
```

Reply:
And delete the rest lol

Reply:
Also why is this a separate function? Why not put this inside `parse_modules/1`? I think that would still be fine.